services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java

```diff
@@ -1139,10 +1145,35 @@ public void ignore() {
                 }
             };
 
+            evaluateApplicationInitiatedActionTrigger(session, provider, model, authSession);
             provider.evaluateTriggers(result);
         }
     }
+    
+    // Determine if provider is being requested as an Application-Initiated Action
+    // If so, add it to the authSession.
+    private static void evaluateApplicationInitiatedActionTrigger(final KeycloakSession session,
+                                                                  final RequiredActionProvider provider, 
+                                                                  final RequiredActionProviderModel model,
+                                                                  final AuthenticationSessionModel authSession
+                                                                  ) {
+        if (provider.initiatedActionSupport() == InitiatedActionSupport.NOT_SUPPORTED) return;
+        
+        String aia = authSession.getClientNote(AIA_REQUEST);
+        if (aia == null) return;
 
+        // make sure you are evaluating the action that was requested
+        if (!aia.equalsIgnoreCase(model.getProviderId())) return;
+        
+        if (session.getContext().getClient().getRole(AccountRoles.INITIATE_ACTION) == null) {
+            logger.error("Client must have initiate-action role to perform application-initiated action.");
+            return;
+        }
+
+        authSession.addRequiredAction(model.getProviderId());
+        authSession.removeClientNote(AIA_REQUEST); // keep this from being executed twice
```

Comment:
It's not clear to me where are you setting `AIA_REQUEST`.

Reply:
I was digging into the sources and got the same impression. @ssilvert when `AIA_REQUEST`is being added to client note?

Reply:
AIA_REQUEST is set by the code that processes the HTTP request.  All extra query params are set as client notes with the LOGIN_SESSION_NOTE_ADDITIONAL_REQ_PARAMS_PREFIX.  So I didn't have to write code for that.

Reply:
private static final String AIA_REQUEST = LOGIN_SESSION_NOTE_ADDITIONAL_REQ_PARAMS_PREFIX + Constants.KC_ACTION;